framework/include/base/SystemBase.h

```diff
@@ -512,8 +514,71 @@ class SystemTempl : public SystemBase
   virtual NumericVector<Number> & solutionUDot() { return *_dummy_vec; }
   virtual Number & duDotDu() { return _du_dot_du; }
 
-  virtual bool hasVector(std::string name) { return _sys.have_vector(name); }
-  virtual NumericVector<Number> & getVector(std::string name) { return _sys.get_vector(name); }
+  /**
+   * Check if the named vector exists in the system.
+   */
+  virtual bool hasVector(const std::string & name) { return _sys.have_vector(name); }
+
+  /**
+   * Get a raw NumericVector with the given name.
+   */
+  virtual NumericVector<Number> & getVector(const std::string & name) { return _sys.get_vector(name); }
+
+  /**
+   * Adds a solution length vector to the system.
+   *
+   * @param vector_name The name of the vector.
+   * @param project Whether or not to project this vector when doing mesh refinement.
+   *                If the vector is just going to be recomputed then there is no need to project it.
+   * @param type What type of parallel vector.  This is usually either PARALLEL or GHOSTED.
+   *                                            GHOSTED is needed if you are going to be accessing off-processor entries.
+   *                                            The ghosting pattern is the same as the solution vector.
+   */
+  virtual NumericVector<Number>& addVector(const std::string & vector_name, const bool project, const ParallelType type)
```

Comment:
Added a new function in `NonlinearSystem`. I realized that heat conduction module is using that capability.